### DISCUSSION TITLE
zrle decoding error on 2nd  frame

### DISCUSSION THREAD

**Discussion by nocanstillbb** [any1/wayvnc#355] [General]
### Useful information:
Please, try to gather as much of useful information as possible and follow
these instructions:

i am try to make a vnc client by myself.    raw format very slow,  so i try to use zrle ,
i have decoded first frame , but the second frame alawys crash,  
i want to know why the second frame of the same image has about 320 bytes less than the first frame

logs 
```
Width:  1280
Height:  720
Bits per pixel:  32
Depth:  24
Big Endian:  0
True Color:  1
Red Max:  255
Green Max:  255
Blue Max:  255
Red Shift:  0
Green Shift:  8
Blue Shift:  16
Name :  "WayVNC"
encodingType: 16   xpos: 0   ypos: 0   width: 1280   height: 720
zrle length : 53167     // ***decoded well.***
encodingType: 16   xpos: 0   ypos: 0   width: 1280   height: 720
zrle length : 52808   // ***2nd  is about 320 bytes less than  1st frame***
ret: -3                       // *** new zlib stream instance inflate return 3, means data error***

```

- **Version:**
  - Run this command: `wayvnc -V`
```bash
dv@rsap5:~ $ wayvnc --version
wayvnc: v0.8.0-f7afe12 (pios)
neatvnc: 0.8.0
aml: 0.3.0
```

- Describe how to **reproduce** the problem

**Comment by any1** [comment]
Is this a crash in your client or in wayvnc? If it's a crash in wayvnc, please try with the newest version (0.9.1) and provide a backtrace if it crashes.

**Comment by nocanstillbb** [comment]
> Is this a crash in your client or in wayvnc? If it's a crash in wayvnc, please try with the newest version (0.9.1) and provide a backtrace if it crashes.

wayvnc hasn't crash,  i just curious why the 2nd frame data can't  uncompress

**Reply by any1** [reply]
The VNC server sends incremental updates. I.e. it only sends that part of the image that changed. Also, the z_stream needs to be preserved for the duration of the session.

This contains all the information that you need: https://github.com/rfbproto/rfbproto/blob/master/rfbproto.rst

**Reply by nocanstillbb** [reply]
I referred to the decoding methods of tiger vnc and realvnc, and I can decode the first frame and draw it perfectly. The number of rectangles, xpos, ypos, width, height, and length of the second frame are also parsed successfully. However, when the data read out according to length is put into the zlib stream, it always returns -3 during inflate. I don't have other vncservers that support no authentication (1) or standard vnc password authentication (2) to test. Realvnc does not support standard vnc password authentication (2), and it seems that no authentication (1) cannot be used.

**Comment by nocanstillbb** [comment]
![image](https://github.com/user-attachments/assets/c80878ec-49f5-4d23-884a-edfc3e929d4d)

**Comment by nocanstillbb** [comment]
It seems that zlib compressed data usually starts with `78..` I compared the data of the first frame and the second frame, and the zrle data of the second frame does not start with 78

**Comment by nocanstillbb** [comment]
wayvnc --log-level=trace
```
Nov 29 22:06:16 rsap5 sh[41053]: Info: New client connection from ::ffff:10.2.1.53: 0x55564df4b430 (ref 1)
Nov 29 22:06:16 rsap5 sh[41053]: DEBUG: ../src/server.c: 845: Client chose security type: 1
Nov 29 22:06:16 rsap5 sh[41053]: Info: Starting screen capture
Nov 29 22:06:16 rsap5 sh[41053]: DEBUG: ../src/main.c: 994: Acquired power state management. Waiting for power event to start capturing
Nov 29 22:06:16 rsap5 sh[41053]: DEBUG: ../src/main.c: 1347: Client connected, new client count: 1
Nov 29 22:06:16 rsap5 sh[41053]: DEBUG: ../src/ctl-server.c: 941: Enqueueing client-connected event: {"id":"37","address":"::","username":null,"seat":"default","connection_count":1}
Nov 29 22:06:16 rsap5 sh[41053]: DEBUG: ../src/ctl-server.c: 968: Enqueued client-connected event for 1 clients
Nov 29 22:06:16 rsap5 sh[41053]: DEBUG: ../src/server.c: 1099: Client 0x55564df4b430 set encodings: zrle,trle,hextile,raw
Nov 29 22:06:16 rsap5 sh[41053]: DEBUG: ../src/server.c: 1013: Using color palette for client 0x55564df4b430
Nov 29 22:06:16 rsap5 sh[41053]: DEBUG: ../src/server.c: 1028: Client 0x55564df4b430 chose pixel format: XRGB8888
Nov 29 22:06:16 rsap5 sh[41053]: Info: Choosing zrle encoding for client 0x55564df4b430
Nov 29 22:06:16 rsap5 sh[41053]: DEBUG: ../src/ctl-server.c: 604: >> {"method":"client-connected","params":{"id":"37","address":"::","username":null,"seat":"default","connection_count":1}}
Nov 29 22:06:16 rsap5 sh[41053]: DEBUG: ../src/main.c: 1003: screencopy_start_immediate

my vnc client app crash

Nov 29 22:06:41 rsap5 sh[41053]: Info: Client 0x55564df4b430 (1) hung up
Nov 29 22:06:41 rsap5 sh[41053]: Info: Closing client connection 0x55564df4b430: ref 0
Nov 29 22:06:41 rsap5 sh[41053]: DEBUG: ../src/main.c: 1294: Client disconnected, new client count: 0
Nov 29 22:06:41 rsap5 sh[41053]: DEBUG: ../src/ctl-server.c: 380: Don't know how to convert sa_family 0 to string
Nov 29 22:06:41 rsap5 sh[41053]: DEBUG: ../src/ctl-server.c: 941: Enqueueing client-disconnected event: {"id":"37","address":null,"username":null,"seat":"default","connection_count":0}
Nov 29 22:06:41 rsap5 sh[41053]: DEBUG: ../src/ctl-server.c: 968: Enqueued client-disconnected event for 1 clients
Nov 29 22:06:41 rsap5 sh[41053]: Info: Stopping screen capture
Nov 29 22:06:41 rsap5 sh[41053]: DEBUG: ../src/ctl-server.c: 604: >> {"method":"client-disconnected","params":{"id":"37","address":null,"username":null,"seat":"default","connection_count":0}}

```

**Comment by nocanstillbb** [comment]
https://stackoverflow.com/questions/76445115/writing-zrle-encoding-for-vnc-server
Finally found the answer here
